src/ring_pedersen.rs

```diff
@@ -0,0 +1,220 @@
+    modulus: BigNumber,
+    /// Ring-Petersen parameter `s`.
+    s: BigNumber,
+    /// Ring-Petersen parameter `t`.
+    t: BigNumber,
+}
+
+/// A commitment produced by [`RingPedersen::commit`].
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub(crate) struct Commitment(BigNumber);
+
+impl Commitment {
+    pub(crate) fn to_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes()
+    }
+}
+
+/// The randomness generated as part of [`RingPedersen::commit`].
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct CommitmentRandomness(BigNumber);
+
+impl CommitmentRandomness {
+    pub(crate) fn to_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes()
+    }
+
+    /// Masks value with `mask` and `challenge`.
+    /// The output [`MaskedRandomness`] value is computed as `mask - challenge * value`.
+    pub(crate) fn mask_neg(
+        &self,
+        mask: &CommitmentRandomness,
+        challenge: &BigNumber,
+    ) -> MaskedRandomness {
+        MaskedRandomness(&mask.0 - challenge * &self.0)
+    }
+
+    /// Masks value with `mask` and `challenge`.
+    /// The output [`MaskedRandomness`] value is computed as `mask + challenge * value`.
+    pub(crate) fn mask(
+        &self,
+        mask: &CommitmentRandomness,
+        challenge: &BigNumber,
+    ) -> MaskedRandomness {
+        MaskedRandomness(&mask.0 + challenge * &self.0)
+    }
+}
+
+/// The randomness generated as part of [`RingPedersen::commit`] and masked
+/// by [`CommitmentRandomness::mask`] or [`CommitmentRandomness::mask_neg`].
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct MaskedRandomness(BigNumber);
+
+impl From<CommitmentRandomness> for MaskedRandomness {
+    fn from(value: CommitmentRandomness) -> Self {
+        Self(value.0)
+    }
+}
+
+impl From<MaskedRandomness> for CommitmentRandomness {
+    fn from(value: MaskedRandomness) -> Self {
+        Self(value.0)
+    }
+}
+
+impl RingPedersen {
+    /// Returns the underlying modulus.
+    pub(crate) fn modulus(&self) -> &BigNumber {
+        &self.modulus
+    }
+
+    pub(crate) fn s(&self) -> &BigNumber {
+        &self.s
+    }
+
+    pub(crate) fn t(&self) -> &BigNumber {
+        &self.t
+    }
+
+    #[cfg(test)]
+    pub(crate) fn build_from_components(s: BigNumber, t: BigNumber, modulus: BigNumber) -> Self {
+        Self { modulus, s, t }
+    }
+
+    /// Extracts a [`RingPedersen`] object from a [`DecryptionKey`] `sk` and
+    /// secret primes `p` and `q`, alongside a zero knowledge proof [`PiPrmProof`]
+    /// that the produced object is validly constructed.
+    ///
+    /// # Panics
+    /// Panics if `p * q` does not equal the modulus of `sk`.
+    pub(crate) fn extract(
+        sk: &DecryptionKey,
+        p: &BigNumber,
+        q: &BigNumber,
+        rng: &mut (impl RngCore + CryptoRng),
+    ) -> Result<(Self, PiPrmProof)> {
+        let ek = sk.encryption_key();
+        let modulus = ek.n();
+        assert_eq!(modulus, &(p * q));
```

Comment:
Can do! My take in the past has been if the error is a mistake in using the API (namely, something that should never happen), it's a panic (since no correct program should ever hit that condition), and if it's a recoverable error in any way, it's an error. I'm fine with returning `InternalInvariantFaield` though, just wanted to explain the reasoning here to see if it's something we'd want to adopt at some point :-).

Reply:
Since we're running this as a library in production, we need to have all our errors be recoverable, even when it's a bug on our end. @gatoWololo  is working on an updated scheme for passing a limited set of actionable errors to the caller and logging all the useful context for the developer. But in the end, if we panic here, we might cause panics up some massive chain of production servers and we can't really do that responsibly.

Reply:
Makes sense!